{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f39793c3-e3f9-4dd1-baaf-bf82443987e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "CondaValueError: prefix already exists: /home/kbanerj1/.conda/envs/JupyterLab\n",
      "\n",
      "Requirement already satisfied: ipykernel in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (6.29.5)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (0.2.2)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (1.8.5)\n",
      "Requirement already satisfied: ipython>=7.23.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (8.18.1)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (8.6.2)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (5.7.2)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (0.1.7)\n",
      "Requirement already satisfied: nest-asyncio in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (1.6.0)\n",
      "Requirement already satisfied: packaging in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (24.1)\n",
      "Requirement already satisfied: psutil in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (6.0.0)\n",
      "Requirement already satisfied: pyzmq>=24 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (26.2.0)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (6.4.1)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipykernel) (5.14.3)\n",
      "Requirement already satisfied: decorator in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (0.19.1)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (3.0.47)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (2.18.0)\n",
      "Requirement already satisfied: stack-data in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (0.6.3)\n",
      "Requirement already satisfied: typing-extensions in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (4.12.2)\n",
      "Requirement already satisfied: exceptiongroup in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (1.2.2)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from ipython>=7.23.1->ipykernel) (4.9.0)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-client>=6.1.12->ipykernel) (8.5.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.9.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (4.3.3)\n",
      "Requirement already satisfied: zipp>=3.20 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from importlib-metadata>=4.8.3->jupyter-client>=6.1.12->ipykernel) (3.20.2)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel) (0.8.4)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from pexpect>4.3->ipython>=7.23.1->ipykernel) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from prompt-toolkit<3.1.0,>=3.0.41->ipython>=7.23.1->ipykernel) (0.2.13)\n",
      "Requirement already satisfied: six>=1.5 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.16.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.1.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /home/kbanerj1/.conda/envs/JupyterLab/lib/python3.9/site-packages (from stack-data->ipython>=7.23.1->ipykernel) (0.2.3)\n",
      "Installed kernelspec JupyterLab in /home/kbanerj1/.local/share/jupyter/kernels/jupyterlab\n"
     ]
    }
   ],
   "source": [
    "!conda env create -f environment.yml\n",
    "!pip install ipykernel\n",
    "!python -m ipykernel install --user --name JupyterLab --display-name \"Python (JupyterLab)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "236bbba3-8faf-443a-a59f-0bf2f99fa357",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pds\n",
    "import glob\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fff16a37-85d6-4a0f-8231-7afce4a9aa7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current Directory: /home/kbanerj1/TCGA_ML\n"
     ]
    }
   ],
   "source": [
    "os.chdir(\"/home/kbanerj1/TCGA_ML\")\n",
    "# Verify the current directory\n",
    "print(\"Current Directory:\", os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e77c57fe-45cd-4a91-b4e9-415f8298a08d",
   "metadata": {},
   "source": [
    "## Reading Gene Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9b533363-49c6-4a8e-8192-b4f1ff3441af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gene_df(filepath, filename):\n",
    "    '''\n",
    "    Function that reads a gene file and processes the gene data with the file name.\n",
    "    Skips rows with unwanted gene names (e.g., N_unmapped, N_multimapping, N_noFeature, N_ambiguous).\n",
    "    Retains only the tpm_unstranded column.\n",
    "    '''\n",
    "    \n",
    "    df1 = pd.read_csv(filepath, sep='\\t', skiprows=1)\n",
    "    unwanted_genes = ['N_unmapped', 'N_multimapping', 'N_noFeature', 'N_ambiguous']\n",
    "    df1 = df1[~df1['gene_id'].isin(unwanted_genes)]\n",
    "    df1 = df1[['gene_name', 'tpm_unstranded']]\n",
    "    df1.set_index('gene_name', inplace=True)\n",
    "    df1 = df1.T\n",
    "    processed_filename = filename.split('.')[0]\n",
    "    df1.index = [processed_filename]\n",
    "    df1 = df1.reset_index()\n",
    "    df1.rename(columns={'index': 'file_name'}, inplace=True)\n",
    "    \n",
    "    return df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e7cb2bb4-d618-4bcc-9bc2-675524961225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>file_name</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>...</th>\n",
       "      <th>AL451106.1</th>\n",
       "      <th>AC092910.4</th>\n",
       "      <th>AC073611.1</th>\n",
       "      <th>AC136977.1</th>\n",
       "      <th>AC078856.1</th>\n",
       "      <th>AC008763.4</th>\n",
       "      <th>AL592295.6</th>\n",
       "      <th>AC006486.3</th>\n",
       "      <th>AL391628.1</th>\n",
       "      <th>AP006621.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0ece26e3-b25e-4b16-b0f8-6a62d128b084</td>\n",
       "      <td>50.8266</td>\n",
       "      <td>0.0394</td>\n",
       "      <td>68.0495</td>\n",
       "      <td>8.9287</td>\n",
       "      <td>1.9557</td>\n",
       "      <td>7.0591</td>\n",
       "      <td>4.9134</td>\n",
       "      <td>2.8817</td>\n",
       "      <td>11.1888</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3160</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.8409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>4.9789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fd950816-2bc3-4e25-90b4-6a6a049cd809</td>\n",
       "      <td>43.9793</td>\n",
       "      <td>0.1093</td>\n",
       "      <td>66.7390</td>\n",
       "      <td>4.1823</td>\n",
       "      <td>2.8904</td>\n",
       "      <td>19.2646</td>\n",
       "      <td>12.2063</td>\n",
       "      <td>66.9193</td>\n",
       "      <td>24.1118</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5660</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.3709</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0335</td>\n",
       "      <td>1.8239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b8d729a4-658d-49c7-b6c1-3490041ad3e1</td>\n",
       "      <td>53.2076</td>\n",
       "      <td>0.1048</td>\n",
       "      <td>54.1207</td>\n",
       "      <td>5.8853</td>\n",
       "      <td>4.1955</td>\n",
       "      <td>5.0059</td>\n",
       "      <td>2.5866</td>\n",
       "      <td>11.0124</td>\n",
       "      <td>31.2974</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.3106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0402</td>\n",
       "      <td>1.7946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>deacae6a-5f74-4a07-8ee7-5f378fa5ea41</td>\n",
       "      <td>57.5719</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>86.4358</td>\n",
       "      <td>4.7965</td>\n",
       "      <td>3.3714</td>\n",
       "      <td>3.2967</td>\n",
       "      <td>3.0484</td>\n",
       "      <td>9.9543</td>\n",
       "      <td>12.3831</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.4321</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>6.4643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36df1baa-1477-4192-a6f1-51585f29f6b2</td>\n",
       "      <td>26.5473</td>\n",
       "      <td>0.0861</td>\n",
       "      <td>61.5650</td>\n",
       "      <td>5.0289</td>\n",
       "      <td>2.1703</td>\n",
       "      <td>7.7559</td>\n",
       "      <td>9.7178</td>\n",
       "      <td>4.3662</td>\n",
       "      <td>15.7490</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4569</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.3981</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0462</td>\n",
       "      <td>7.1829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>edf2bc5d-3b50-411a-819e-aac3aa1207aa</td>\n",
       "      <td>67.4699</td>\n",
       "      <td>1.6255</td>\n",
       "      <td>78.4940</td>\n",
       "      <td>7.7254</td>\n",
       "      <td>4.3927</td>\n",
       "      <td>13.4459</td>\n",
       "      <td>8.2017</td>\n",
       "      <td>29.0845</td>\n",
       "      <td>30.7405</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.2797</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.1102</td>\n",
       "      <td>1.2953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>ee33c2fd-140b-4b4c-8820-f711facc6b4c</td>\n",
       "      <td>78.6831</td>\n",
       "      <td>1.8529</td>\n",
       "      <td>60.1268</td>\n",
       "      <td>6.2156</td>\n",
       "      <td>5.1865</td>\n",
       "      <td>8.4333</td>\n",
       "      <td>21.7751</td>\n",
       "      <td>45.8958</td>\n",
       "      <td>15.1308</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.4865</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0812</td>\n",
       "      <td>1.3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>814e81ff-bb80-4c3c-bd94-d86c89de9a36</td>\n",
       "      <td>9.4880</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>36.2045</td>\n",
       "      <td>2.9239</td>\n",
       "      <td>3.0229</td>\n",
       "      <td>1.2153</td>\n",
       "      <td>49.4383</td>\n",
       "      <td>33.9834</td>\n",
       "      <td>7.3017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.1495</td>\n",
       "      <td>0.1907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.7446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3007</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>1.2654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>200bfb25-663b-4216-be51-bafaea7c5d4a</td>\n",
       "      <td>37.7357</td>\n",
       "      <td>0.0758</td>\n",
       "      <td>245.2148</td>\n",
       "      <td>6.3458</td>\n",
       "      <td>12.5770</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>68.7707</td>\n",
       "      <td>9.0761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.9733</td>\n",
       "      <td>0.7199</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.7262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.9586</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>0.7993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>4078cbdc-60c2-4d55-a2de-b497f3ee1a6e</td>\n",
       "      <td>49.3302</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>52.6570</td>\n",
       "      <td>4.3309</td>\n",
       "      <td>2.0570</td>\n",
       "      <td>2.0234</td>\n",
       "      <td>0.8106</td>\n",
       "      <td>15.3136</td>\n",
       "      <td>35.1777</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4277</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.5665</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>2.4551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>707 rows × 60661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name                             file_name   TSPAN6    TNMD      DPM1  \\\n",
       "0          0ece26e3-b25e-4b16-b0f8-6a62d128b084  50.8266  0.0394   68.0495   \n",
       "1          fd950816-2bc3-4e25-90b4-6a6a049cd809  43.9793  0.1093   66.7390   \n",
       "2          b8d729a4-658d-49c7-b6c1-3490041ad3e1  53.2076  0.1048   54.1207   \n",
       "3          deacae6a-5f74-4a07-8ee7-5f378fa5ea41  57.5719  0.0000   86.4358   \n",
       "4          36df1baa-1477-4192-a6f1-51585f29f6b2  26.5473  0.0861   61.5650   \n",
       "..                                          ...      ...     ...       ...   \n",
       "702        edf2bc5d-3b50-411a-819e-aac3aa1207aa  67.4699  1.6255   78.4940   \n",
       "703        ee33c2fd-140b-4b4c-8820-f711facc6b4c  78.6831  1.8529   60.1268   \n",
       "704        814e81ff-bb80-4c3c-bd94-d86c89de9a36   9.4880  0.0000   36.2045   \n",
       "705        200bfb25-663b-4216-be51-bafaea7c5d4a  37.7357  0.0758  245.2148   \n",
       "706        4078cbdc-60c2-4d55-a2de-b497f3ee1a6e  49.3302  0.0000   52.6570   \n",
       "\n",
       "gene_name   SCYL3  C1orf112      FGR      CFH    FUCA2     GCLC  ...  \\\n",
       "0          8.9287    1.9557   7.0591   4.9134   2.8817  11.1888  ...   \n",
       "1          4.1823    2.8904  19.2646  12.2063  66.9193  24.1118  ...   \n",
       "2          5.8853    4.1955   5.0059   2.5866  11.0124  31.2974  ...   \n",
       "3          4.7965    3.3714   3.2967   3.0484   9.9543  12.3831  ...   \n",
       "4          5.0289    2.1703   7.7559   9.7178   4.3662  15.7490  ...   \n",
       "..            ...       ...      ...      ...      ...      ...  ...   \n",
       "702        7.7254    4.3927  13.4459   8.2017  29.0845  30.7405  ...   \n",
       "703        6.2156    5.1865   8.4333  21.7751  45.8958  15.1308  ...   \n",
       "704        2.9239    3.0229   1.2153  49.4383  33.9834   7.3017  ...   \n",
       "705        6.3458   12.5770   0.0000   0.0000  68.7707   9.0761  ...   \n",
       "706        4.3309    2.0570   2.0234   0.8106  15.3136  35.1777  ...   \n",
       "\n",
       "gene_name  AL451106.1  AC092910.4  AC073611.1  AC136977.1  AC078856.1  \\\n",
       "0                 0.0      0.0000      0.3160         0.0      0.0000   \n",
       "1                 0.0      0.0000      0.5660         0.0      0.0000   \n",
       "2                 0.0      0.0000      0.6104         0.0      0.0000   \n",
       "3                 0.0      0.0000      0.2795         0.0      0.0000   \n",
       "4                 0.0      0.0000      0.4569         0.0      0.0000   \n",
       "..                ...         ...         ...         ...         ...   \n",
       "702               0.0      0.0000      0.5189         0.0      0.0000   \n",
       "703               0.0      0.0000      0.3599         0.0      0.0000   \n",
       "704               0.0      3.1495      0.1907         0.0      7.7446   \n",
       "705               0.0      4.9733      0.7199         0.0      6.7262   \n",
       "706               0.0      0.0000      0.4277         0.0      0.0000   \n",
       "\n",
       "gene_name  AC008763.4  AL592295.6  AC006486.3  AL391628.1  AP006621.6  \n",
       "0                 0.0     17.8409         0.0      0.0483      4.9789  \n",
       "1                 0.0      9.3709         0.0      0.0335      1.8239  \n",
       "2                 0.0     26.3106         0.0      0.0402      1.7946  \n",
       "3                 0.0     19.4321         0.0      0.0151      6.4643  \n",
       "4                 0.0     17.3981         0.0      0.0462      7.1829  \n",
       "..                ...         ...         ...         ...         ...  \n",
       "702               0.0     18.2797         0.0      0.1102      1.2953  \n",
       "703               0.0     16.4865         0.0      0.0812      1.3372  \n",
       "704               0.0      2.3007         0.0      0.0540      1.2654  \n",
       "705               0.0      8.9586         0.0      0.0349      0.7993  \n",
       "706               0.0     22.5665         0.0      0.0390      2.4551  \n",
       "\n",
       "[707 rows x 60661 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the list of file paths\n",
    "filelist = glob.glob(\"./data/*/*.tsv\")\n",
    "all_dfs = []\n",
    "\n",
    "for filepath in filelist:\n",
    "    filename =  os.path.basename(filepath)  # get filename from path\n",
    "    df = read_gene_df(filepath, filename)\n",
    "    all_dfs.append(df) \n",
    "\n",
    "combined_df = pd.concat(all_dfs, ignore_index=True)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3363615b-7e6c-4a2d-9b7a-4e50c45d9239",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CD99', 'NME1-NME2', 'Z83844.1', 'MCTS2P', 'CORO7-PAM16', 'ICAM4',\n",
       "       'AC004837.1', 'AL021546.1', 'PEDS1-UBE2V1', 'VAMP7',\n",
       "       ...\n",
       "       'CDR1', 'AC114982.2', 'AC114402.2', 'AC084756.2', 'AL031178.2',\n",
       "       'ACTL10', 'AC119733.1', 'AL451106.1', 'AC136977.1', 'AC006486.3'],\n",
       "      dtype='object', name='gene_name', length=2657)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## sanity check for columns that have all zeros across rows (samples)\n",
    "zero_read_genes = combined_df.columns[(combined_df == 0).all(axis=0)]\n",
    "zero_read_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bee8a87b-340c-4f7a-afae-e38997116e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>file_name</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>...</th>\n",
       "      <th>AC010980.1</th>\n",
       "      <th>AC007511.1</th>\n",
       "      <th>AC004233.4</th>\n",
       "      <th>AC092910.4</th>\n",
       "      <th>AC073611.1</th>\n",
       "      <th>AC078856.1</th>\n",
       "      <th>AC008763.4</th>\n",
       "      <th>AL592295.6</th>\n",
       "      <th>AL391628.1</th>\n",
       "      <th>AP006621.6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0ece26e3-b25e-4b16-b0f8-6a62d128b084</td>\n",
       "      <td>50.8266</td>\n",
       "      <td>0.0394</td>\n",
       "      <td>68.0495</td>\n",
       "      <td>8.9287</td>\n",
       "      <td>1.9557</td>\n",
       "      <td>7.0591</td>\n",
       "      <td>4.9134</td>\n",
       "      <td>2.8817</td>\n",
       "      <td>11.1888</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1370</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3995</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3160</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.8409</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>4.9789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fd950816-2bc3-4e25-90b4-6a6a049cd809</td>\n",
       "      <td>43.9793</td>\n",
       "      <td>0.1093</td>\n",
       "      <td>66.7390</td>\n",
       "      <td>4.1823</td>\n",
       "      <td>2.8904</td>\n",
       "      <td>19.2646</td>\n",
       "      <td>12.2063</td>\n",
       "      <td>66.9193</td>\n",
       "      <td>24.1118</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8968</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1109</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5660</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.3709</td>\n",
       "      <td>0.0335</td>\n",
       "      <td>1.8239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b8d729a4-658d-49c7-b6c1-3490041ad3e1</td>\n",
       "      <td>53.2076</td>\n",
       "      <td>0.1048</td>\n",
       "      <td>54.1207</td>\n",
       "      <td>5.8853</td>\n",
       "      <td>4.1955</td>\n",
       "      <td>5.0059</td>\n",
       "      <td>2.5866</td>\n",
       "      <td>11.0124</td>\n",
       "      <td>31.2974</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4690</td>\n",
       "      <td>0.1319</td>\n",
       "      <td>0.1063</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.6104</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.3106</td>\n",
       "      <td>0.0402</td>\n",
       "      <td>1.7946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>deacae6a-5f74-4a07-8ee7-5f378fa5ea41</td>\n",
       "      <td>57.5719</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>86.4358</td>\n",
       "      <td>4.7965</td>\n",
       "      <td>3.3714</td>\n",
       "      <td>3.2967</td>\n",
       "      <td>3.0484</td>\n",
       "      <td>9.9543</td>\n",
       "      <td>12.3831</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2689</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1992</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2795</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.4321</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>6.4643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36df1baa-1477-4192-a6f1-51585f29f6b2</td>\n",
       "      <td>26.5473</td>\n",
       "      <td>0.0861</td>\n",
       "      <td>61.5650</td>\n",
       "      <td>5.0289</td>\n",
       "      <td>2.1703</td>\n",
       "      <td>7.7559</td>\n",
       "      <td>9.7178</td>\n",
       "      <td>4.3662</td>\n",
       "      <td>15.7490</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6208</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3494</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4569</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.3981</td>\n",
       "      <td>0.0462</td>\n",
       "      <td>7.1829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>edf2bc5d-3b50-411a-819e-aac3aa1207aa</td>\n",
       "      <td>67.4699</td>\n",
       "      <td>1.6255</td>\n",
       "      <td>78.4940</td>\n",
       "      <td>7.7254</td>\n",
       "      <td>4.3927</td>\n",
       "      <td>13.4459</td>\n",
       "      <td>8.2017</td>\n",
       "      <td>29.0845</td>\n",
       "      <td>30.7405</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0752</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2302</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5189</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.2797</td>\n",
       "      <td>0.1102</td>\n",
       "      <td>1.2953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>ee33c2fd-140b-4b4c-8820-f711facc6b4c</td>\n",
       "      <td>78.6831</td>\n",
       "      <td>1.8529</td>\n",
       "      <td>60.1268</td>\n",
       "      <td>6.2156</td>\n",
       "      <td>5.1865</td>\n",
       "      <td>8.4333</td>\n",
       "      <td>21.7751</td>\n",
       "      <td>45.8958</td>\n",
       "      <td>15.1308</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3950</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3599</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.4865</td>\n",
       "      <td>0.0812</td>\n",
       "      <td>1.3372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>814e81ff-bb80-4c3c-bd94-d86c89de9a36</td>\n",
       "      <td>9.4880</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>36.2045</td>\n",
       "      <td>2.9239</td>\n",
       "      <td>3.0229</td>\n",
       "      <td>1.2153</td>\n",
       "      <td>49.4383</td>\n",
       "      <td>33.9834</td>\n",
       "      <td>7.3017</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0796</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.1495</td>\n",
       "      <td>0.1907</td>\n",
       "      <td>7.7446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.3007</td>\n",
       "      <td>0.0540</td>\n",
       "      <td>1.2654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>200bfb25-663b-4216-be51-bafaea7c5d4a</td>\n",
       "      <td>37.7357</td>\n",
       "      <td>0.0758</td>\n",
       "      <td>245.2148</td>\n",
       "      <td>6.3458</td>\n",
       "      <td>12.5770</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>68.7707</td>\n",
       "      <td>9.0761</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1131</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2052</td>\n",
       "      <td>4.9733</td>\n",
       "      <td>0.7199</td>\n",
       "      <td>6.7262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.9586</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>0.7993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>4078cbdc-60c2-4d55-a2de-b497f3ee1a6e</td>\n",
       "      <td>49.3302</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>52.6570</td>\n",
       "      <td>4.3309</td>\n",
       "      <td>2.0570</td>\n",
       "      <td>2.0234</td>\n",
       "      <td>0.8106</td>\n",
       "      <td>15.3136</td>\n",
       "      <td>35.1777</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3159</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4277</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.5665</td>\n",
       "      <td>0.0390</td>\n",
       "      <td>2.4551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>707 rows × 56941 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name                             file_name   TSPAN6    TNMD      DPM1  \\\n",
       "0          0ece26e3-b25e-4b16-b0f8-6a62d128b084  50.8266  0.0394   68.0495   \n",
       "1          fd950816-2bc3-4e25-90b4-6a6a049cd809  43.9793  0.1093   66.7390   \n",
       "2          b8d729a4-658d-49c7-b6c1-3490041ad3e1  53.2076  0.1048   54.1207   \n",
       "3          deacae6a-5f74-4a07-8ee7-5f378fa5ea41  57.5719  0.0000   86.4358   \n",
       "4          36df1baa-1477-4192-a6f1-51585f29f6b2  26.5473  0.0861   61.5650   \n",
       "..                                          ...      ...     ...       ...   \n",
       "702        edf2bc5d-3b50-411a-819e-aac3aa1207aa  67.4699  1.6255   78.4940   \n",
       "703        ee33c2fd-140b-4b4c-8820-f711facc6b4c  78.6831  1.8529   60.1268   \n",
       "704        814e81ff-bb80-4c3c-bd94-d86c89de9a36   9.4880  0.0000   36.2045   \n",
       "705        200bfb25-663b-4216-be51-bafaea7c5d4a  37.7357  0.0758  245.2148   \n",
       "706        4078cbdc-60c2-4d55-a2de-b497f3ee1a6e  49.3302  0.0000   52.6570   \n",
       "\n",
       "gene_name   SCYL3  C1orf112      FGR      CFH    FUCA2     GCLC  ...  \\\n",
       "0          8.9287    1.9557   7.0591   4.9134   2.8817  11.1888  ...   \n",
       "1          4.1823    2.8904  19.2646  12.2063  66.9193  24.1118  ...   \n",
       "2          5.8853    4.1955   5.0059   2.5866  11.0124  31.2974  ...   \n",
       "3          4.7965    3.3714   3.2967   3.0484   9.9543  12.3831  ...   \n",
       "4          5.0289    2.1703   7.7559   9.7178   4.3662  15.7490  ...   \n",
       "..            ...       ...      ...      ...      ...      ...  ...   \n",
       "702        7.7254    4.3927  13.4459   8.2017  29.0845  30.7405  ...   \n",
       "703        6.2156    5.1865   8.4333  21.7751  45.8958  15.1308  ...   \n",
       "704        2.9239    3.0229   1.2153  49.4383  33.9834   7.3017  ...   \n",
       "705        6.3458   12.5770   0.0000   0.0000  68.7707   9.0761  ...   \n",
       "706        4.3309    2.0570   2.0234   0.8106  15.3136  35.1777  ...   \n",
       "\n",
       "gene_name  AC010980.1  AC007511.1  AC004233.4  AC092910.4  AC073611.1  \\\n",
       "0              0.1370      0.0000      0.3995      0.0000      0.3160   \n",
       "1              0.8968      0.0000      0.1109      0.0000      0.5660   \n",
       "2              0.4690      0.1319      0.1063      0.0000      0.6104   \n",
       "3              1.2689      0.0000      0.1992      0.0000      0.2795   \n",
       "4              0.6208      0.0000      0.3494      0.0000      0.4569   \n",
       "..                ...         ...         ...         ...         ...   \n",
       "702            0.0752      0.0000      0.2302      0.0000      0.5189   \n",
       "703            0.3950      0.0000      0.0000      0.0000      0.3599   \n",
       "704            0.0796      0.0000      0.0000      3.1495      0.1907   \n",
       "705            0.1131      0.0000      0.2052      4.9733      0.7199   \n",
       "706            0.3159      0.0000      0.0000      0.0000      0.4277   \n",
       "\n",
       "gene_name  AC078856.1  AC008763.4  AL592295.6  AL391628.1  AP006621.6  \n",
       "0              0.0000         0.0     17.8409      0.0483      4.9789  \n",
       "1              0.0000         0.0      9.3709      0.0335      1.8239  \n",
       "2              0.0000         0.0     26.3106      0.0402      1.7946  \n",
       "3              0.0000         0.0     19.4321      0.0151      6.4643  \n",
       "4              0.0000         0.0     17.3981      0.0462      7.1829  \n",
       "..                ...         ...         ...         ...         ...  \n",
       "702            0.0000         0.0     18.2797      0.1102      1.2953  \n",
       "703            0.0000         0.0     16.4865      0.0812      1.3372  \n",
       "704            7.7446         0.0      2.3007      0.0540      1.2654  \n",
       "705            6.7262         0.0      8.9586      0.0349      0.7993  \n",
       "706            0.0000         0.0     22.5665      0.0390      2.4551  \n",
       "\n",
       "[707 rows x 56941 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop zero count genes\n",
    "combined_df = combined_df.drop(columns=zero_read_genes)\n",
    "combined_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353d47c2-89fd-463d-ae2b-471338f7d7c3",
   "metadata": {},
   "source": [
    "## Read clinical and sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "141d1dec-6c6b-4927-ba9b-cd4e2659d46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>project_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>age_is_obfuscated</th>\n",
       "      <th>cause_of_death</th>\n",
       "      <th>cause_of_death_source</th>\n",
       "      <th>country_of_residence_at_enrollment</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>...</th>\n",
       "      <th>treatment_arm</th>\n",
       "      <th>treatment_dose</th>\n",
       "      <th>treatment_dose_units</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_effect_indicator</th>\n",
       "      <th>treatment_frequency</th>\n",
       "      <th>treatment_intent_type</th>\n",
       "      <th>treatment_or_therapy</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>treatment_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Immunotherapy (Including Vaccines)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Targeted Molecular Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Adjuvant</td>\n",
       "      <td>yes</td>\n",
       "      <td>'--</td>\n",
       "      <td>Chemotherapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d420e653-3fb2-432b-9e81-81232a80264d</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>'--</td>\n",
       "      <td>False</td>\n",
       "      <td>Cancer Related</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>-19586</td>\n",
       "      <td>481</td>\n",
       "      <td>...</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "      <td>Neoadjuvant</td>\n",
       "      <td>no</td>\n",
       "      <td>'--</td>\n",
       "      <td>'--</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 158 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id  case_submitter_id project_id  \\\n",
       "0  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "1  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "2  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "3  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "4  d420e653-3fb2-432b-9e81-81232a80264d  HCM-BROD-0210-C71  HCMI-CMDC   \n",
       "\n",
       "  age_at_index age_is_obfuscated  cause_of_death cause_of_death_source  \\\n",
       "0          '--             False  Cancer Related                   '--   \n",
       "1          '--             False  Cancer Related                   '--   \n",
       "2          '--             False  Cancer Related                   '--   \n",
       "3          '--             False  Cancer Related                   '--   \n",
       "4          '--             False  Cancer Related                   '--   \n",
       "\n",
       "  country_of_residence_at_enrollment days_to_birth days_to_death  ...  \\\n",
       "0                                '--        -19586           481  ...   \n",
       "1                                '--        -19586           481  ...   \n",
       "2                                '--        -19586           481  ...   \n",
       "3                                '--        -19586           481  ...   \n",
       "4                                '--        -19586           481  ...   \n",
       "\n",
       "  treatment_arm treatment_dose treatment_dose_units treatment_effect  \\\n",
       "0           '--            '--                  '--              '--   \n",
       "1           '--            '--                  '--              '--   \n",
       "2           '--            '--                  '--              '--   \n",
       "3           '--            '--                  '--              '--   \n",
       "4           '--            '--                  '--              '--   \n",
       "\n",
       "  treatment_effect_indicator treatment_frequency treatment_intent_type  \\\n",
       "0                        '--                 '--              Adjuvant   \n",
       "1                        '--                 '--              Adjuvant   \n",
       "2                        '--                 '--              Adjuvant   \n",
       "3                        '--                 '--              Adjuvant   \n",
       "4                        '--                 '--           Neoadjuvant   \n",
       "\n",
       "   treatment_or_therapy treatment_outcome                      treatment_type  \n",
       "0                    no               '--  Immunotherapy (Including Vaccines)  \n",
       "1                    no               '--          Targeted Molecular Therapy  \n",
       "2                    no               '--              Radiation Therapy, NOS  \n",
       "3                   yes               '--                        Chemotherapy  \n",
       "4                    no               '--                                 '--  \n",
       "\n",
       "[5 rows x 158 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load clinical data \n",
    "clinical_data = pd.read_csv(\"./metadata/clinical.tsv\", sep='\\t')\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ee7956e-5102-463a-b969-457267b78123",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1386, 158)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cd294ea7-0118-4f98-8acd-fe4e40cef8f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "primary_diagnosis\n",
       "Astrocytoma, NOS                 118\n",
       "Astrocytoma, anaplastic          240\n",
       "Glioblastoma                     434\n",
       "Gliosarcoma                        6\n",
       "Mixed glioma                     242\n",
       "Oligodendroglioma, NOS           202\n",
       "Oligodendroglioma, anaplastic    144\n",
       "Name: case_id, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data.groupby(\"primary_diagnosis\")['case_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9328141a-0624-4d93-b904-0721dfc7dc35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(663,\n",
       " case_submitter_id\n",
       " HCM-BROD-0213-C71    8\n",
       " HCM-BROD-0104-C71    7\n",
       " HCM-BROD-0106-C71    7\n",
       " HCM-BROD-0420-C71    7\n",
       " HCM-BROD-0198-C71    7\n",
       " Name: count, dtype: int64)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_cases_in_clinical = clinical_data['case_submitter_id'].nunique()\n",
    "case_duplicates = clinical_data['case_submitter_id'].value_counts()\n",
    "# Display the unique case count and the distribution of duplicates\n",
    "unique_cases_in_clinical, case_duplicates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "450596e9-f792-402d-bc78-dc3ae9ae44c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(663, 158)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop duplicates based on 'case_submitter_id'\n",
    "clinical_data_dedup = clinical_data.drop_duplicates(subset='case_submitter_id', keep='first')\n",
    "clinical_data_dedup.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c257052e-2716-483e-847a-92535e090890",
   "metadata": {},
   "source": [
    "## Simplify labels to broader categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37a0b644-2fba-4fb9-8a5c-30d360ca782d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3072523/80191726.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  clinical_data_dedup.loc[:, 'label'] = clinical_data_dedup['primary_diagnosis'].map(diagnosis_map)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "primary_diagnosis              label            \n",
       "Astrocytoma, NOS               Astrocytoma           59\n",
       "Astrocytoma, anaplastic        Astrocytoma          120\n",
       "Glioblastoma                   Glioblastoma         189\n",
       "Oligodendroglioma, NOS         Oligodendroglioma    101\n",
       "Oligodendroglioma, anaplastic  Oligodendroglioma     72\n",
       "Name: case_submitter_id, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# simplify labels by mapping complex or specific diagnoses to broader categories - dropping 'mixed gliomas'\n",
    "diagnosis_map = {\n",
    "    \"Astrocytoma, NOS\": \"Astrocytoma\",\n",
    "    \"Astrocytoma, anaplastic\": \"Astrocytoma\",\n",
    "    \"Oligodendroglioma, NOS\": \"Oligodendroglioma\",\n",
    "    \"Oligodendroglioma, anaplastic\": \"Oligodendroglioma\",\n",
    "    \"Glioblastoma\": \"Glioblastoma\"\n",
    "}\n",
    "\n",
    "# create a new 'label' column based on the mapping\n",
    "clinical_data_dedup.loc[:, 'label'] = clinical_data_dedup['primary_diagnosis'].map(diagnosis_map)\n",
    "clinical_data_dedup.groupby([\"primary_diagnosis\", \"label\"])['case_submitter_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "97f7826e-6b31-4c2d-8fdb-abce4e0824c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>filename_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>92cdbe72-8af4-4e98-87be-9c437978eb0c</td>\n",
       "      <td>2fb43643-4dc3-41b3-a299-815f5c8b0d72.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03748</td>\n",
       "      <td>C3L-03748-04</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>2fb43643-4dc3-41b3-a299-815f5c8b0d72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>99e99371-68ea-49e0-adea-cd59a61e4539</td>\n",
       "      <td>502a976d-ce43-4b4a-b813-58eaa2b0c387.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>HCM-BROD-0106-C71-02A</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "      <td>502a976d-ce43-4b4a-b813-58eaa2b0c387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a3f23052-28f1-44b9-b2e5-54ca544f16ed</td>\n",
       "      <td>8f11a23c-f4da-462f-95e9-4c85e56f603a.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>HCM-BROD-0106-C71-85A</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>8f11a23c-f4da-462f-95e9-4c85e56f603a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dee27d65-9700-4d94-bd5d-8d67a34747a0</td>\n",
       "      <td>8d272434-7ba5-4287-9f1f-f4fac4e1fc7b.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0103-C71</td>\n",
       "      <td>HCM-BROD-0103-C71-85B</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>8d272434-7ba5-4287-9f1f-f4fac4e1fc7b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3f5c7c90-3c15-46fd-8d33-a8cff36325c4</td>\n",
       "      <td>014c0152-04c3-4370-9fc6-e42bdc7f3f79.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCM-BROD-0210-C71-02A</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "      <td>014c0152-04c3-4370-9fc6-e42bdc7f3f79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                File ID  \\\n",
       "0  92cdbe72-8af4-4e98-87be-9c437978eb0c   \n",
       "1  99e99371-68ea-49e0-adea-cd59a61e4539   \n",
       "2  a3f23052-28f1-44b9-b2e5-54ca544f16ed   \n",
       "3  dee27d65-9700-4d94-bd5d-8d67a34747a0   \n",
       "4  3f5c7c90-3c15-46fd-8d33-a8cff36325c4   \n",
       "\n",
       "                                           File Name            Data Category  \\\n",
       "0  2fb43643-4dc3-41b3-a299-815f5c8b0d72.rna_seq.a...  Transcriptome Profiling   \n",
       "1  502a976d-ce43-4b4a-b813-58eaa2b0c387.rna_seq.a...  Transcriptome Profiling   \n",
       "2  8f11a23c-f4da-462f-95e9-4c85e56f603a.rna_seq.a...  Transcriptome Profiling   \n",
       "3  8d272434-7ba5-4287-9f1f-f4fac4e1fc7b.rna_seq.a...  Transcriptome Profiling   \n",
       "4  014c0152-04c3-4370-9fc6-e42bdc7f3f79.rna_seq.a...  Transcriptome Profiling   \n",
       "\n",
       "                        Data Type Project ID            Case ID  \\\n",
       "0  Gene Expression Quantification    CPTAC-3          C3L-03748   \n",
       "1  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0106-C71   \n",
       "2  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0106-C71   \n",
       "3  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0103-C71   \n",
       "4  Gene Expression Quantification  HCMI-CMDC  HCM-BROD-0210-C71   \n",
       "\n",
       "               Sample ID                   Sample Type  \\\n",
       "0           C3L-03748-04                 Primary Tumor   \n",
       "1  HCM-BROD-0106-C71-02A               Recurrent Tumor   \n",
       "2  HCM-BROD-0106-C71-85A  Next Generation Cancer Model   \n",
       "3  HCM-BROD-0103-C71-85B  Next Generation Cancer Model   \n",
       "4  HCM-BROD-0210-C71-02A               Recurrent Tumor   \n",
       "\n",
       "                         filename_short  \n",
       "0  2fb43643-4dc3-41b3-a299-815f5c8b0d72  \n",
       "1  502a976d-ce43-4b4a-b813-58eaa2b0c387  \n",
       "2  8f11a23c-f4da-462f-95e9-4c85e56f603a  \n",
       "3  8d272434-7ba5-4287-9f1f-f4fac4e1fc7b  \n",
       "4  014c0152-04c3-4370-9fc6-e42bdc7f3f79  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_data = pd.read_csv(\"./metadata/gdc_sample_sheet.tsv\", sep='\\t')\n",
    "# Simplify filenames in sample_data for merging\n",
    "sample_data['filename_short'] = sample_data['File Name'].apply(lambda x: x.split('.')[0])\n",
    "sample_data = pd.merge(sample_data, combined_df, left_on='filename_short', right_on='file_name', how='inner')\n",
    "\n",
    "# Retain relevant columns in sample_data\n",
    "columns_to_retain = ['File ID', 'File Name', 'Data Category', 'Data Type', 'Project ID', 'Case ID', 'Sample ID', 'Sample Type', 'filename_short']\n",
    "sample_data = sample_data[columns_to_retain]\n",
    "sample_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a4342ff7-b5f7-4042-a734-ad4714fda691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(707, 9)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "37c31ebc-6d67-42e6-8c0e-84977263294a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File ID</th>\n",
       "      <th>File Name</th>\n",
       "      <th>Data Category</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Case ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>filename_short</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>92cdbe72-8af4-4e98-87be-9c437978eb0c</td>\n",
       "      <td>2fb43643-4dc3-41b3-a299-815f5c8b0d72.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03748</td>\n",
       "      <td>C3L-03748-04</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>2fb43643-4dc3-41b3-a299-815f5c8b0d72</td>\n",
       "      <td>C3L-03748</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>99e99371-68ea-49e0-adea-cd59a61e4539</td>\n",
       "      <td>502a976d-ce43-4b4a-b813-58eaa2b0c387.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>HCM-BROD-0106-C71-02A</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "      <td>502a976d-ce43-4b4a-b813-58eaa2b0c387</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a3f23052-28f1-44b9-b2e5-54ca544f16ed</td>\n",
       "      <td>8f11a23c-f4da-462f-95e9-4c85e56f603a.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>HCM-BROD-0106-C71-85A</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>8f11a23c-f4da-462f-95e9-4c85e56f603a</td>\n",
       "      <td>HCM-BROD-0106-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dee27d65-9700-4d94-bd5d-8d67a34747a0</td>\n",
       "      <td>8d272434-7ba5-4287-9f1f-f4fac4e1fc7b.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0103-C71</td>\n",
       "      <td>HCM-BROD-0103-C71-85B</td>\n",
       "      <td>Next Generation Cancer Model</td>\n",
       "      <td>8d272434-7ba5-4287-9f1f-f4fac4e1fc7b</td>\n",
       "      <td>HCM-BROD-0103-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3f5c7c90-3c15-46fd-8d33-a8cff36325c4</td>\n",
       "      <td>014c0152-04c3-4370-9fc6-e42bdc7f3f79.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>HCM-BROD-0210-C71-02A</td>\n",
       "      <td>Recurrent Tumor</td>\n",
       "      <td>014c0152-04c3-4370-9fc6-e42bdc7f3f79</td>\n",
       "      <td>HCM-BROD-0210-C71</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>73d57ab1-1757-4af3-b229-90050dcdb9e0</td>\n",
       "      <td>1fbd55b4-c08b-44b8-940e-6538cc79ef68.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>C3L-00104-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>1fbd55b4-c08b-44b8-940e-6538cc79ef68</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>0545b5c3-1dfb-4e0e-b956-97ba018a1406</td>\n",
       "      <td>9b4b6b42-09e1-4783-b374-532b9df2e207.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03681</td>\n",
       "      <td>C3L-03681-03</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>9b4b6b42-09e1-4783-b374-532b9df2e207</td>\n",
       "      <td>C3L-03681</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>Glioblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>c46154dd-81dd-4203-960c-da74e393eccd</td>\n",
       "      <td>cb6d1cb9-0b4b-4938-902b-39d02b07faf7.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-02705, C3L-02705, C3L-02705</td>\n",
       "      <td>C3L-02705-71, C3L-02705-75, C3L-02705-72</td>\n",
       "      <td>Primary Tumor, Primary Tumor, Primary Tumor</td>\n",
       "      <td>cb6d1cb9-0b4b-4938-902b-39d02b07faf7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>04a95ed6-a66b-4a95-9b12-da3b48ab7ff6</td>\n",
       "      <td>0ab80f6a-8139-42a9-886c-e8eb5cc9de88.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-02704, C3L-02704, C3L-02704</td>\n",
       "      <td>C3L-02704-71, C3L-02704-73, C3L-02704-72</td>\n",
       "      <td>Primary Tumor, Primary Tumor, Primary Tumor</td>\n",
       "      <td>0ab80f6a-8139-42a9-886c-e8eb5cc9de88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>953318ce-676f-4a47-bca9-6f618ce71382</td>\n",
       "      <td>02383b35-4baa-4875-9127-fcdd2707977c.rna_seq.a...</td>\n",
       "      <td>Transcriptome Profiling</td>\n",
       "      <td>Gene Expression Quantification</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>C3L-03744, C3L-03744, C3L-03744</td>\n",
       "      <td>C3L-03744-71, C3L-03744-75, C3L-03744-72</td>\n",
       "      <td>Primary Tumor, Primary Tumor, Primary Tumor</td>\n",
       "      <td>02383b35-4baa-4875-9127-fcdd2707977c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>707 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  File ID  \\\n",
       "0    92cdbe72-8af4-4e98-87be-9c437978eb0c   \n",
       "1    99e99371-68ea-49e0-adea-cd59a61e4539   \n",
       "2    a3f23052-28f1-44b9-b2e5-54ca544f16ed   \n",
       "3    dee27d65-9700-4d94-bd5d-8d67a34747a0   \n",
       "4    3f5c7c90-3c15-46fd-8d33-a8cff36325c4   \n",
       "..                                    ...   \n",
       "702  73d57ab1-1757-4af3-b229-90050dcdb9e0   \n",
       "703  0545b5c3-1dfb-4e0e-b956-97ba018a1406   \n",
       "704  c46154dd-81dd-4203-960c-da74e393eccd   \n",
       "705  04a95ed6-a66b-4a95-9b12-da3b48ab7ff6   \n",
       "706  953318ce-676f-4a47-bca9-6f618ce71382   \n",
       "\n",
       "                                             File Name  \\\n",
       "0    2fb43643-4dc3-41b3-a299-815f5c8b0d72.rna_seq.a...   \n",
       "1    502a976d-ce43-4b4a-b813-58eaa2b0c387.rna_seq.a...   \n",
       "2    8f11a23c-f4da-462f-95e9-4c85e56f603a.rna_seq.a...   \n",
       "3    8d272434-7ba5-4287-9f1f-f4fac4e1fc7b.rna_seq.a...   \n",
       "4    014c0152-04c3-4370-9fc6-e42bdc7f3f79.rna_seq.a...   \n",
       "..                                                 ...   \n",
       "702  1fbd55b4-c08b-44b8-940e-6538cc79ef68.rna_seq.a...   \n",
       "703  9b4b6b42-09e1-4783-b374-532b9df2e207.rna_seq.a...   \n",
       "704  cb6d1cb9-0b4b-4938-902b-39d02b07faf7.rna_seq.a...   \n",
       "705  0ab80f6a-8139-42a9-886c-e8eb5cc9de88.rna_seq.a...   \n",
       "706  02383b35-4baa-4875-9127-fcdd2707977c.rna_seq.a...   \n",
       "\n",
       "               Data Category                       Data Type Project ID  \\\n",
       "0    Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "1    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "2    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "3    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "4    Transcriptome Profiling  Gene Expression Quantification  HCMI-CMDC   \n",
       "..                       ...                             ...        ...   \n",
       "702  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "703  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "704  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "705  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "706  Transcriptome Profiling  Gene Expression Quantification    CPTAC-3   \n",
       "\n",
       "                             Case ID  \\\n",
       "0                          C3L-03748   \n",
       "1                  HCM-BROD-0106-C71   \n",
       "2                  HCM-BROD-0106-C71   \n",
       "3                  HCM-BROD-0103-C71   \n",
       "4                  HCM-BROD-0210-C71   \n",
       "..                               ...   \n",
       "702                        C3L-00104   \n",
       "703                        C3L-03681   \n",
       "704  C3L-02705, C3L-02705, C3L-02705   \n",
       "705  C3L-02704, C3L-02704, C3L-02704   \n",
       "706  C3L-03744, C3L-03744, C3L-03744   \n",
       "\n",
       "                                    Sample ID  \\\n",
       "0                                C3L-03748-04   \n",
       "1                       HCM-BROD-0106-C71-02A   \n",
       "2                       HCM-BROD-0106-C71-85A   \n",
       "3                       HCM-BROD-0103-C71-85B   \n",
       "4                       HCM-BROD-0210-C71-02A   \n",
       "..                                        ...   \n",
       "702                              C3L-00104-01   \n",
       "703                              C3L-03681-03   \n",
       "704  C3L-02705-71, C3L-02705-75, C3L-02705-72   \n",
       "705  C3L-02704-71, C3L-02704-73, C3L-02704-72   \n",
       "706  C3L-03744-71, C3L-03744-75, C3L-03744-72   \n",
       "\n",
       "                                     Sample Type  \\\n",
       "0                                  Primary Tumor   \n",
       "1                                Recurrent Tumor   \n",
       "2                   Next Generation Cancer Model   \n",
       "3                   Next Generation Cancer Model   \n",
       "4                                Recurrent Tumor   \n",
       "..                                           ...   \n",
       "702                                Primary Tumor   \n",
       "703                                Primary Tumor   \n",
       "704  Primary Tumor, Primary Tumor, Primary Tumor   \n",
       "705  Primary Tumor, Primary Tumor, Primary Tumor   \n",
       "706  Primary Tumor, Primary Tumor, Primary Tumor   \n",
       "\n",
       "                           filename_short  case_submitter_id  \\\n",
       "0    2fb43643-4dc3-41b3-a299-815f5c8b0d72          C3L-03748   \n",
       "1    502a976d-ce43-4b4a-b813-58eaa2b0c387  HCM-BROD-0106-C71   \n",
       "2    8f11a23c-f4da-462f-95e9-4c85e56f603a  HCM-BROD-0106-C71   \n",
       "3    8d272434-7ba5-4287-9f1f-f4fac4e1fc7b  HCM-BROD-0103-C71   \n",
       "4    014c0152-04c3-4370-9fc6-e42bdc7f3f79  HCM-BROD-0210-C71   \n",
       "..                                    ...                ...   \n",
       "702  1fbd55b4-c08b-44b8-940e-6538cc79ef68          C3L-00104   \n",
       "703  9b4b6b42-09e1-4783-b374-532b9df2e207          C3L-03681   \n",
       "704  cb6d1cb9-0b4b-4938-902b-39d02b07faf7                NaN   \n",
       "705  0ab80f6a-8139-42a9-886c-e8eb5cc9de88                NaN   \n",
       "706  02383b35-4baa-4875-9127-fcdd2707977c                NaN   \n",
       "\n",
       "    primary_diagnosis         label  \n",
       "0        Glioblastoma  Glioblastoma  \n",
       "1        Glioblastoma  Glioblastoma  \n",
       "2        Glioblastoma  Glioblastoma  \n",
       "3        Glioblastoma  Glioblastoma  \n",
       "4        Glioblastoma  Glioblastoma  \n",
       "..                ...           ...  \n",
       "702      Glioblastoma  Glioblastoma  \n",
       "703      Glioblastoma  Glioblastoma  \n",
       "704               NaN           NaN  \n",
       "705               NaN           NaN  \n",
       "706               NaN           NaN  \n",
       "\n",
       "[707 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with the clinical data to get the primary diagnosis and label\n",
    "merged_data_dedup = sample_data.merge(\n",
    "    clinical_data_dedup[['case_submitter_id', 'primary_diagnosis', 'label']], \n",
    "    how='left', \n",
    "    left_on='Case ID', \n",
    "    right_on='case_submitter_id'\n",
    ")\n",
    "\n",
    "merged_data_dedup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "17f777f0-6ffb-44fb-a0d0-b77cce82a0c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(707, 12)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data_dedup.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fa87b0a8-553a-4073-a0c5-ea1eea06caf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "primary_diagnosis              label            \n",
       "Astrocytoma, NOS               Astrocytoma           61\n",
       "Astrocytoma, anaplastic        Astrocytoma          121\n",
       "Glioblastoma                   Glioblastoma         209\n",
       "Oligodendroglioma, NOS         Oligodendroglioma    108\n",
       "Oligodendroglioma, anaplastic  Oligodendroglioma     76\n",
       "Name: case_submitter_id, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by both primary_diagnosis and label, then count unique case_submitter_ids\n",
    "merged_data_dedup.groupby(['primary_diagnosis', 'label'])['case_submitter_id'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9a2fbf5c-bf9f-4ec2-9962-8329ebc34f92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(575, 12)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter out samples with non-empty labels\n",
    "merged_data_dedup = merged_data_dedup[merged_data_dedup['label'].notna()]\n",
    "merged_data_dedup.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "489564ba-c065-4acd-90f7-acb2b4a10bbd",
   "metadata": {},
   "source": [
    "## Keep samples that have valid labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1c507332-a705-405a-9b38-15b51cb9e94f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(575, 56940)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge combined_df with merged_data_dedup to get 'Sample ID'\n",
    "combined_df_with_sample_id = pd.merge(combined_df, merged_data_dedup[['filename_short', 'Sample ID']], \n",
    "    left_on='file_name', right_on='filename_short', how='left')\n",
    "\n",
    "# Set 'Sample ID' as index, drop unnecessary columns\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.set_index('Sample ID')\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.drop(['filename_short', 'file_name'], axis=1)\n",
    "\n",
    "# Filter combined_df_with_sample_id to only include non-empty labels\n",
    "filtered_sample_ids = merged_data_dedup['Sample ID'].dropna().unique()\n",
    "combined_df_with_sample_id = combined_df_with_sample_id.loc[filtered_sample_ids]\n",
    "\n",
    "# Display the final shape\n",
    "combined_df_with_sample_id.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57eceaa5-dfff-46b9-ab74-a3cc297d6db4",
   "metadata": {},
   "source": [
    "## Output the Processed Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7c3fd380-374d-4e49-935b-391bebf49bff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data saved to ./processed/data.csv\n"
     ]
    }
   ],
   "source": [
    "# Define extraction directory\n",
    "extraction_dir = \"./processed\"\n",
    "os.makedirs(extraction_dir, exist_ok=True)\n",
    "\n",
    "data_file_path = os.path.join(extraction_dir, 'data.csv')\n",
    "combined_df_with_sample_id.to_csv(data_file_path, index=True)\n",
    "print(f\"Data saved to {data_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f88a9f5f-8429-4d0a-bc58-0378c38697a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Labels saved to ./processed/glioma_labels.csv\n"
     ]
    }
   ],
   "source": [
    "# Extract and save the filtered labels\n",
    "y = merged_data_dedup.set_index('Sample ID')['label']\n",
    "labels_file_path = os.path.join(extraction_dir, 'glioma_labels.csv')\n",
    "y.to_csv(labels_file_path, index=True)\n",
    "print(f\"Labels saved to {labels_file_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (JupyterLab)",
   "language": "python",
   "name": "jupyterlab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
